src/dmd/dscope.d

```diff
@@ -799,4 +798,16 @@ struct Scope
         }
         return false;
     }
+    /**
+     * dmd relies on mutation of state during semantic analysis, however
+     * sometimes semantic is being performed in a speculative context that should
+     * not have any visible effect on the rest of the compilation: for example when compiling
+     * a typeof() or __traits(compiles).
+     *
+     * Returns: `true` if this `Scope` is known to be from one of these speculative contexts
+     */
+    extern(C++) bool isFromSpeculativeSemanticContext() scope
```

Comment:
A detail but the name is really long, why not `isSpeculative`?

Reply:
Also, why C++? Does ldc use this function?

Reply:
Why not C++? The second part I cannot answer causally

Reply:
Keep internal semantic machinery internal.

Reply:
It isn't machinery as per se, this could be useful for culling emitted symbols or more weakly a speculative symbol may still need to be emitted into the binary but doesn't need to be optimized because it'll never be called.

Reply:
That's still internal to the front-end (see `dmd.dtemplate.needsCodegen`).